gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/access/operations/SortedMap2IndexInsertOperation.java

```diff
@@ -287,18 +290,47 @@ else if (lockTimeout > 0) { // wait before throwing conflict
 
         // TODO: identify the cause of Bug SNAP-2627
         // Till then this is the crude fix
-        if (oldValue instanceof AbstractRegionEntry &&
-            ((AbstractRegionEntry)oldValue).isDestroyedOrRemoved()) {
-          // create a dummy exception
-          Throwable th = GemFireXDUtils.newDuplicateKeyViolation("unique constraint",
-            container.getQualifiedTableName(), "key=" + key.toString()
-            + ", row=" + value, oldValue, null, null);
-          final GemFireCacheImpl cache = Misc.getGemFireCache();
-          final LogWriter logger = cache.getLogger();
-          logger.error("Unique index constraint violation caused due to " +
-            "removed/destroyed entry. The index is corrupted. Cleaning the index", th);
-          skipListMap.remove(key, oldValue);
-          continue;
+        if (oldValue instanceof AbstractRegionEntry) {
+          final AbstractRegionEntry existingRe = (AbstractRegionEntry)oldValue;
+          if (existingRe.isDestroyedOrRemoved()) {
+            Callable<Boolean> deadEntryRemover = new Callable<Boolean>() {
+              @Override
+              public Boolean call() {
+                synchronized (existingRe) {
+                  if (existingRe.isDestroyedOrRemoved()) {
```

Comment:
Not sure how using a separate thread avoids a deadlock. This thread will be waiting on this thread to end, no?

Lets say the problem is lock inversion namely (newRe, existingRe). The problem this separate thread is probably trying to get around is lock order the other way (existingRe, newRe). Now for the first, existingRe is a separate thread but that will be waited for in this thread so lock ordering will remain the same. Its a timed wait so this can go into a retry but the lock ordering (newRe, existingRe) will remain the same in retries too.

Reply:
Yeah seems fine. I missed the fact that on receiving an exception (including timeout), the code will throw constraint violation and not continue in the loop.